lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,859 @@
+      Agent.update(agent1, &Map.put(&1, :key, "value"))
+      Agent.get(agent1, & &1)
+      #=> %{key: "value"}
+
+      {:ok, agent2} = DynamicSupervisor.start_link(sup, {Agent, fn -> %{} end})
+      Agent.get(agent2, & &1)
+      #=> %{}
+
+      DynamicSupervisor.count_children(sup_pid)
+      #=> %{active: 2, specs: 2, supervisors: 0, workers: 2}
+
+  ## Module-based supervisors
+
+  Similar to `Supervisor`, dynamic supervisors also support module-based
+  supervisors.
+
+      defmodule MyApp.DynamicSupervisor do
+        # Automatically defines child_spec/1
+        use DynamicSupervisor
+
+        def start_link(arg) do
+          DynamicSupervisor.start_link(__MODULE__, arg, name: __MODULE__)
+        end
+
+        def init(_arg) do
+          DynamicSupervisor.init(strategy: :one_for_one)
+        end
+      end
+
+  See the `Supervisor` docs for a discussion of when you may want to use
+  module-based supervisors.
+
+  ## Name registration
+
+  A supervisor is bound to the same name registration rules as a `GenServer`.
+  Read more about these rules in the documentation for `GenServer`.
+  """
+
+  @behaviour GenServer
+
+  @doc """
+  Callback invoked to start the supervisor and during hot code upgrades.
+  """
+  @callback init(args :: term) :: {:ok, sup_flags()} | :ignore
+
+  @typep sup_flags() :: %{
+           strategy: strategy(),
+           intensity: non_neg_integer(),
+           period: pos_integer(),
+           max_children: non_neg_integer()
+         }
+
+  @typedoc "Option values used by the `start*` functions"
+  @type option :: {:name, Supervisor.name()} | flag()
+
+  @typedoc "Options used by the `start*` functions"
+  @type options :: [option, ...]
+
+  @typedoc "Options given to `start_link/2` and `init/2`"
+  @type flag ::
+          {:strategy, strategy}
+          | {:max_restarts, non_neg_integer}
+          | {:max_seconds, pos_integer}
+          | {:max_children, non_neg_integer | :infinity}
+
+  @typedoc "Supported strategies"
+  @type strategy :: :one_for_one
+
+  defstruct [
+    :args,
+    :mod,
+    :name,
+    :strategy,
+    :max_restarts,
+    :max_seconds,
+    :max_children,
+    children: %{},
+    restarts: [],
+    dynamic: 0
+  ]
+
+  @doc false
+  defmacro __using__(opts) do
+    quote location: :keep do
+      @behaviour DynamicSupervisor
+      @opts unquote(opts)
+
+      @doc false
+      def child_spec(arg) do
+        default = %{
+          id: __MODULE__,
+          start: {__MODULE__, :start_link, [arg]},
+          type: :supervisor
+        }
+
+        Supervisor.child_spec(default, @opts)
+      end
+
+      defoverridable child_spec: 1
+
+      @doc false
+      def init(arg)
+    end
+  end
+
+  @doc """
+  Starts a supervisor with the given options.
+
+  The `:strategy` is a required option and the currently supported
+  value is `:one_for_one`. The remaining options can be found in the
+  `init/1` docs.
```

Comment:
Maybe explain these options here instead, and refer to this function in `init/1`. I'd expect we'll most often use this supervisor without the callback module.